component/kafka/group/component.go

```diff
@@ -167,13 +167,19 @@ type Component struct {
 	batchTimeout time.Duration
 	retries      uint
 	retryWait    time.Duration
+	topicChecker func(ctx context.Context, c *Component) error
 	commitSync   bool
 }
 
 // Run starts the consumer processing loop to process messages from Kafka.
 func (c *Component) Run(ctx context.Context) error {
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
+	if c.topicChecker != nil {
```

Comment:
If topicChecker is set through a `OptionFunc`, we run it before starting processing. The goal is to have the patron `Service` to fail at startup in case those 'unrecoverable' errors verify. Notice that the `topicChecker` is opt-out by default for backwards compatibility.

Reply:
There is an quirk with this approach.
The component Run is executed when we call the service run of patron.
This could mean that other component might have already started.
This means that if this fails , patron will cancel all the others also which might have strange side effects.
With the discussed approach we would not even go to patron service in order to run it it would have failed much earlier.
WDYT?

Reply:
I see what you mean. I thought about that too - I did not put at first into the constructor to make sure to raise the discussion. I'm ok with both approaches but would like to have the support of the maintainers in whichever decision is deemed the preferred:

1. Execute the `topicChecker` in `Run`:
    * (pro) Network connection/call made in the same scope as before - preserve of the `lazy` approach when doing network connection only when needed;
    * (pro) Usage of the same cancelable `context.Context` used to do the real processing;
    * (con) Might interfere with other components of the `Service` because would cancel the context of the whole service. 
 2. Execute the `topicChecker` by the end of the `New` constructor:
     * (pro) Less invasive to the rest of the `Component` sharing the same `Service`
     * (con) Need to either take a `context.Context` in the constructor `New` or in the `CheckTopic` method, which might be a different context than the one used by the Run method
     * (con) A constructor that establishes a network connection instead of the previous `lazy` connection approach.
     
Those were the tradeoffs I could gather, depending on how we weight those and how we add more concerns to the tradeoff we can choose the preferred method. Again, I'm ok with both approaches.

Reply:
What about the initial OptionFunc discussion?

Reply:
I have summarized my comprehension about that discussion in the issue - [see the comment](https://github.com/beatlabs/patron/issues/443#issuecomment-996012466). 

I might have missed something... 

Let me see if I understand your interpretation:

You thought that the idea of using `OptionFunc` was exactly to be able to call the method in the constructor in an 'Opt-in' fashion?

Reply:
[Update] - The `pro` of reusing the `context.Context` is not really applicable because as @teivah spot, it is not really needed to make the two calls we perform.

Reply:
So the tradeoff is quite simple after all:

1. Execute the `topicChecker` in `Run`:
    * (pro) Network connection/call only made lazily at the `Service` `Run` 
    * (con) When failing it will interfere with all the other components -> less flexibility to a client, which might want to continue with Service startup even if one component fails...
2. Execute `topicChecker` by the end of the New constructor:
    * (pro) less invasive to the resto of the `Component`s sharing the same `Service` -> more flexibility to the patron clients which might want to proceed with the `Service` with a subset of `Component`s
    * (con) a constructor that establishes a network connection differs behavior-wise from what the other components make. This is nevertheless mitigated by the fact t hat the behavior is Opt-in
    
I'm also basically 'sold' for the approach number 2 as you suggested. Are we in the same page? Still all convinced that solution 2 is better? @teivah and @mantzas ?

Reply:
The initial discussion was for the code to live in the option func and not part of component.
The option func will be executed in the loop in whatever order as we have established that it would not matter since the config does not need to be shared. If the check fails the option func will return an error and will be handled in the loop.
Benefits of this approach is that the component does not even know that this exists (follows the open close priciple).
So it is not exactly option 2, more like a clean version of it.
WDYT?

Reply:
The only way we can do this is by not leveraging on the `AllowAutoTopicCreation` that is set for the component. This means that we need to **always** fail if the topic does not exist which can potentially break component tests for instance...

Reply:
Do you see the limitation? Do you have a suggestion on how to circumvent it while still using the execution in the loop?

Reply:
Notice how also in the tests I have covered the two different behaviors:

1. [here](https://github.com/beatlabs/patron/pull/446/files#diff-b51b211c3eb5b3c5c04444579dc1f2eb9e473ef95ae42d12688f7f9490d2d8a0R236)
2. [here](https://github.com/beatlabs/patron/pull/446/files#diff-b51b211c3eb5b3c5c04444579dc1f2eb9e473ef95ae42d12688f7f9490d2d8a0R266)

Reply:
The ability to create topic or not is only enable in our test environment because we allow it from the Kafka broker and not the client. In production you cannot create topics even if the client config says so. In component tests we actually create the topics when the container starts e.g. ` KAFKA_CREATE_TOPICS: "Topic1:1:3,Topic2:1:1:compact"`. I am not sure though about our staging environments.
So setting the client config should not have any effect. This needs to be validated though. 
Can we check this first?

Reply:
I had in mind component tests of patron-clients - e.g. `sonar` where the topics are created by the `producer` (and the producer can reach the local broker after the consumer)

Reply:
But I also agree that since this is a opt-in functionality we might decide to delegate to patron-clients this complexity (i.e, create the topics before starting up the service)

Reply:
If we decide to go for the 'strict' approach (the one that always checks the existence of the topic regardless of the value of `AllowAutoTopicCreation` I would not be worried about staging because we indeed want to fail fast in case the topic is not there in staging. The impact will only be in adapting patron-clients component tests (which currently rely on the fact that the topic will be created automatically).